src/amber/cli/recipes/recipe.cr

```diff
@@ -0,0 +1,148 @@
+require "teeplate"
+require "liquid"
+require "base64"
+require "./filters"
+
+require "random/secure"
+require "../helpers/helpers"
+
+require "./file_entries"
+require "./recipe_fetcher"
+require "./app"
+require "./controller"
+require "./scaffold/controller"
+require "./scaffold/view"
+
+module Amber::Recipes
+  AMBER_RECIPE_FOLDER = "./.amber_recipe_cache"
+
+  class Recipe
+    getter name : String
+    getter directory : String
+    getter recipe : String
+
+    def self.can_generate?(template_type, recipe)
+      return false unless ["app", "controller", "scaffold" ].includes? template_type
+
+      template = RecipeFetcher.new(template_type, recipe).fetch
+      template.nil? ? false : true
+    end
+
+    def initialize(name : String, directory : String, recipe : String, fields = [] of String)
+      if name.match(/\A[a-zA-Z]/)
+        @name = name.underscore
+      else
+        raise "Name is not valid."
+      end
+
+      @directory = File.join(directory)
+      unless Dir.exists?(@directory)
+        Dir.mkdir_p(@directory)
+      end
+
+      @recipe = recipe
+
+      @fields = fields
+    end
+
+    def cleanup
+      # if we used a template from github then remove the temp folder
+      # this should probably be called as a command so templates are cached
+      FileUtils.rm_rf(AMBER_RECIPE_FOLDER)
+    end
+
+    def generate(template : String, options = nil)
+      case template
+      when "app"
+        if options
+          puts "Rendering App #{name} in #{directory} from #{options.r}"
+          App.new(name, options.d, options.t, options.m, options.r).render(directory, list: true, color: true)
+          if options.deps?
+            puts "Installing Dependencies"
+            Amber::CLI::Helpers.run("cd #{name} && shards update")
+          end
+        end
+      when "controller"
+        puts "Rendering Controller #{name} from #{@recipe}"
+        Controller.new(name, @recipe, @fields).render(directory, list: true, color: true)
+      when "scaffold"
+        puts "Rendering Scaffold #{name} from #{@recipe}"
+        if model == "crecto"
+          Amber::CLI::CrectoMigration.new(name, @fields).render(directory, list: true, color: true)
+          Amber::CLI::CrectoModel.new(name, @fields).render(directory, list: true, color: true)
+        else
+          Amber::CLI::GraniteMigration.new(name, @fields).render(directory, list: true, color: true)
+          Amber::CLI::GraniteModel.new(name, @fields).render(directory, list: true, color: true)
+        end
+        Scaffold::Controller.new(name, @recipe, @fields).render(directory, list: true, color: true)
+        Scaffold::View.new(name, @recipe, @fields).render(directory, list: true, color: true)
+      else
+        CLI.logger.error "Template not found", "Generate", :light_red
+      end
+    end
+
+    def model
+      CLI.config.model
+    end
+
+    def puts(msg)
```

Comment:
I missed this on the first time through, but shadowing puts makes me a little uncomfortable. I've done it in the past on some of my projects and regretted it. How about just `log`?